### PR TITLE
[Diffusion][MammothModa2] Enable module-level CPU offload with VAE staging

### PR DESCRIPTION
**Pull request by avicii-forever**
## Purpose

Enable component-level (module-mode) CPU offload for MammothModa2-Preview by
implementing the `SupportsModelCpuOffload` protocol (following the MiniMax-H3
template), so the VAE decoder is staged alongside the DiT and image-condition
refiner.

> **Stacked on #7134** (@Levius-Fubuki shared diffusion runtime migration).
> This branch contains #7134's commits plus the offload changes. The
> **offload-only diff is 2 files**: `pipeline_mammothmoda2_dit.py` and
> `test_mammothmoda2_module_offload.py`.

Refs #7158, #7075.

## Changes (offload-only)

- `_validate_module_offload_runtime()`: reject non-Preview or multi-request
  module offload before building modules; wired into `__init__`.
- `enable_omni_model_cpu_offload()` / `disable_omni_model_cpu_offload()`: stage
  the DiT against the image-condition refiner and VAE decoder, re-activating
  the VAE for its non-forward `decode` entry point via `_component_on_device`.
- Two offload runtime fixes:
  - `model_device = self.device` (not `next(self.parameters()).device`, which
    probes offloaded CPU placeholders).
  - `@torch.no_grad()` (not `inference_mode`, which blocks parameter rebinding).
- 10 CPU contract tests (component discovery, strategy resolution, admission,
  and offload enable/disable lifecycle).

## Validation

NVIDIA A800 80GB, MammothModa2-Preview T2I 1024x1024, 50 steps, CFG 4.0, seed 42.

| | baseline | offload |
| --- | ---: | ---: |
| End-to-end latency | 99.58 s | 100.07 s |
| Peak GPU memory | 50177 MiB | 45405 MiB (-4772 MiB) |
| Output correctness | valid | bit-exact identical |

AI assistance was used in drafting this change.

### VAE in the compact offload selector

  The compact `diffusion_offload_config.components` selector now accepts `vae`
  (module mode only), so `components: ["dit", "vae"]` stages the VAE decoder out
  during denoising and re-activates it for its non-`forward` `decode` entry point
  through the existing `_component_on_device` wrapper — instead of only being
  reachable via the legacy full-topology path.

  - `offloader/config.py`: add `VAE_COMPONENT`, extend `OFFLOAD_COMPONENTS`,
    require `mode="module"` for `vae`, and add VAE to the legacy transfer map.
  - `offloader/base.py`: include VAE in the default `dlo_transfers`.
  - `offloader/plan_resolver.py`: reject `vae` for pipelines that do not own a
    VAE lifecycle (`SupportsModelCpuOffload`).
  - `mammoth_moda2/pipeline_mammothmoda2_dit.py`: select the VAE into
    `offload_encoder_modules` when `vae` is requested.
  - `minimax_h3/pipeline_minimax_h3.py`: reject `vae` in the compact selector —
    MiniMax-H3 stages its VAEs through the compatibility topology only, so the
    selector is not yet wired for `vae` there.

  ### `_parse_request` decomposition

  Split the monolithic `_parse_request` into `_parse_ar_conditions`,
  `_resolve_image_size`, and `_parse_sampling`. Behavior is unchanged — all
  validation messages and the standard-field-over-legacy-alias precedence are
  preserved (also fixes an overlong line flagged by lint).

  ### Tests

  - `test_mammothmoda2_module_offload.py`: compact `vae` selection + missing-VAE guard.
  - `test_mammothmoda2_module_offload_gpu.py`: VAE leaves GPU after decode under `["dit", "vae"]`.
  - `test_minimax_h3_offload.py`: H3 rejects `vae` in the compact selector.
  - `test_layerwise_backend.py`: `vae` requires `mode="module"`; DLO transfer maps updated.

### CODE REVIEW

**Comment by avicii-forever** [comment]
Self-review:

- **Scope**: component-level (module-mode) CPU offload for MammothModa2-Preview, stacked on #7134. The offload-only diff is 2 files (`pipeline_mammothmoda2_dit.py` + `test_mammothmoda2_module_offload.py`).
- **Admission**: `_validate_module_offload_runtime` rejects non-Preview and multi-request configs before model construction; wired into `__init__`.
- **VAE offload**: `SupportsModelCpuOffload` stages the DiT against the refiner + VAE, with `_component_on_device` wrapping the refiner and the non-forward VAE `decode` entry point. Note: VAE is staged via the legacy full-topology path; the compact `components` selector still only covers `dit`/`text_encoder`.
- **Runtime fixes**: `model_device = self.device` (offload replaces param storage with CPU placeholders) and `@torch.no_grad()` (inference_mode blocks param rebinding).
- **Tests**: 10 CPU contract tests + 2 GPU lifecycle tests (real backend, no mocks — enable moves the DiT to CPU, disable restores weights bit-exactly). 40/40 pass with the existing suite.

**Validation** (A800 80GB, Preview T2I, 50 steps, CFG 4.0, seed 42):

Multiple runs (warm, excluding cold start):
- baseline: 103.2–103.6 s (mean 103.4 s), peak 50169 MiB
- offload: 103.7–106.9 s (mean 105.3 s), peak 45397 MiB (-4772 MiB / -9.5%)
- output bit-exact identical to baseline

H2D/D2H breakdown: DiT 5.5 GB load ~200 ms (once), empty_cache ~100 ms per offload; total transfer overhead is small.

Headroom (1536×1536): baseline peak 56521 MiB vs offload 46741 MiB (-9780 MiB / -17.3%). Offload@1536 uses less memory than baseline@1024, confirming capacity headroom.

Known follow-ups: a precise per-transfer CUDA-event breakdown and the cold-start (offload init) peak remain to be characterized.

AI assistance was used in drafting this change.

File: vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py [hunk 1]
```diff
@@ -123,39 +197,109 @@ def _reinit_caption_embedder(self, in_features: int) -> None:
...
+        height, width = dimensions
+        if height <= 0 or width <= 0:
+            raise ValueError(f"Invalid image size: {height}x{width} for request {request_id}")
+        if height % 16 != 0 or width % 16 != 0:
+            raise ValueError(f"Image size must be multiples of 16, got {height}x{width} for request {request_id}")
 
-    def embed_input_ids(self, input_ids: torch.Tensor) -> torch.Tensor:
-        # DiT stage does not consume token embeddings; return a dummy tensor.
+        extra_args = sampling.extra_args or {}
+        guidance = extra_args.get("text_guidance_scale")
```
Comment (BLANKETusers) [inline vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py:255]: [P1] Please prevent deployment defaults from overriding explicitly provided standard sampling fields.

The MammothModa2 stage configuration injects `extra_args.text_guidance_scale=9.0` and `extra_args.num_inference_steps=50` as defaults. This parser gives those legacy aliases priority over `sampling.guidance_scale` and `sampling.num_inference_steps`.

Therefore, a request specifying `guidance_scale=4.0` or `num_inference_steps=20` still runs with 9.0 and 50 unless the caller also provides the legacy aliases through `extra_body`.

Could we remove these aliases from the deployment defaults, or only apply them when they were explicitly supplied by the user?

Comment (avicii-forever) [reply vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py]: Thanks for the catch — confirmed. The stage-1 default_sampling_params.extra_args in deploy/mammoth_moda2.yaml always injects text_guidance_scale=9.0 / num_inference_steps=50, so extra_args.get(...) is never None and the standard guidance_scale / num_inference_steps fields are silently ignored.

  I'll flip the precedence so an explicitly-provided standard field wins and the legacy alias only acts as a fallback:

  guidance = sampling.guidance_scale if sampling.guidance_scale_provided else extra_args.get("text_guidance_scale")
  if guidance is None:
      guidance = 9.0
  ...
  raw_num_inference_steps = sampling.num_inference_steps
  if raw_num_inference_steps is None:
      raw_num_inference_steps = extra_args.get("num_inference_steps")
  if raw_num_inference_steps is None:
      raw_num_inference_steps = 50

  This keeps extra_body / legacy-alias callers working while letting the first-class fields take effect. I'll also update test_parse_request_prefers_legacy_sampling_overrides,which currently asserts the legacy-alias-wins behavior, to assert the standard field wins instead.

Comment (avicii-forever) [reply vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py]: @BLANKETusers thanks for the review — addressed in commit d773114. Summary of the changes in this round:

  1. P1 — sampling precedence (your comment)

  Flipped the precedence in _parse_request so an explicitly-provided standard field wins and the legacy extra_args aliases only act as a fallback:

  guidance = sampling.guidance_scale if sampling.guidance_scale_provided else extra_args.get("text_guidance_scale")
  ...
  raw_num_inference_steps = sampling.num_inference_steps
  if raw_num_inference_steps is None:
      raw_num_inference_steps = extra_args.get("num_inference_steps")

  This way the stage-1 deployment defaults (extra_args.text_guidance_scale=9.0 / num_inference_steps=50) can no longer shadow a caller-supplied guidance_scale / num_inference_steps. Renamed test_parse_request_prefers_legacy_sampling_overrides → test_parse_request_standard_fields_override_legacy_aliases and flipped its assertions accordingly.

  2. CI failure — test_alignment.py

  test_all_non_excluded_pipelines_have_tests failed because MammothModa2DiTPipeline is now registered in _DIFFUSION_MODELS but had no tiny-model test settings. Added it to EXCLUDED_MODELS with a comment (it requires AR-stage hidden states, so it's covered by the dedicated tests/diffusion/models/mammoth_moda2 contract tests instead of a tiny-model builder).

  3. GPU lifecycle test actually running in CI

  test_mammothmoda2_module_offload_gpu.py was marked core_model, diffusion, which is a category tag, not a hardware mark — so it was deselected by both -m 'core_model and cpu' and  -m 'core_model and cuda'. Changed to core_model, cuda so it now runs in the CUDA CI.

File: vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py [hunk 2]
```diff
@@ -389,16 +478,75 @@ def _ensure_2d(x: torch.Tensor, name: str) -> torch.Tensor:
...
+        selection_options: dict[str, Any] = {}
+        if offload_components is not None:
+            if DIT_COMPONENT in offload_components and not dits:
+                raise ValueError("MammothModa2 has no loaded DiT for selected module offload")
+            if TEXT_ENCODER_COMPONENT in offload_components and not components.encoders:
+                raise ValueError("MammothModa2 has no loaded text encoder for selected module offload")
+            selection_options = {
+                "offload_dit_modules": dits if DIT_COMPONENT in offload_components else (),
+                "offload_encoder_modules": (
+                    components.encoders if TEXT_ENCODER_COMPONENT in offload_components else ()
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py]: Include the VAE in compact-config offload targets so it leaves GPU during denoising.

File: vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py [hunk 3]
```diff
@@ -123,39 +197,113 @@ def _reinit_caption_embedder(self, in_features: int) -> None:
             nn.Linear(in_features, out_features, bias=True),
         )
 
-    def get_dummy_runtime_additional_information(self, num_reqs: int) -> list[dict[str, object]]:
-        if num_reqs <= 0:
-            raise ValueError(f"num_reqs must be positive, got {num_reqs}")
-        if num_reqs > 1:
-            raise NotImplementedError(
-                f"get_dummy_runtime_additional_information does not support num_reqs > 1, got {num_reqs}"
+    def _parse_request(self, req: DiffusionRequestBatch) -> _MammothRequest:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py:202]: Split AR-condition validation, image sizing, and sampling parsing into private helpers.

Comment (avicii-forever) [reply vllm_omni/diffusion/models/mammoth_moda2/pipeline_mammothmoda2_dit.py:202]: @lishunyang12 thanks — both points addressed in bf82559.

  1. **VAE in compact-config offload** (`pipeline_mammothmoda2_dit.py:517`): added
     `VAE_COMPONENT` to the offload grammar (`vae`, module mode only) and wired it
     into `enable_omni_model_cpu_offload`, so the VAE is staged out during denoising
     and re-activated through the existing `_component_on_device` wrapper around
     `decode`. Pipelines that don't own a VAE lifecycle now reject `vae` in
     `resolve_offload_plan`, and MiniMax-H3 rejects it too (it still stages its VAEs
     through the compatibility topology only).

  2. **Split `_parse_request`** (`:200`): extracted `_parse_ar_conditions`,
     `_resolve_image_size`, and `_parse_sampling`. Behavior is unchanged — all
     validation messages and the standard-field-over-legacy-alias precedence are
     preserved.

  Tests: compact `vae` selection + missing-VAE guard (CPU), VAE-leaves-GPU-after-decode
  (GPU), MiniMax-H3 `vae` rejection, and `vae` requires `mode="module"`.

**Comment by avicii-forever** [comment]
Responding to the review bot's nudge — here's the current plan and status.

**Rebased onto latest `main`** and dropped the four #7134 commits (that PR merged on 2026-09-18, so they're no longer needed on this branch). The diff is now the offload work only: **14 files, +577/−72**.

**CI** — this PR never received a buildkite run before (the branch had drifted 217 commits behind `main`, and mergeability was UNKNOWN). The rebase re-triggers it now.

**Earlier review feedback** is addressed and carried through the rebase:
- @BLANKETusers's P1 (sampling precedence) — explicit `guidance_scale` / `num_inference_steps` now win over the deploy-injected `extra_args` defaults.
- @lishunyang12's two points — `vae` added to the compact-config offload selector (module mode only), and `_parse_request` split into `_parse_ar_conditions` / `_resolve_image_size` / `_parse_sampling`.

@david6666666 @Isotr0py @princepride — this is ready for a formal review. The 14-file diff is larger than the core two-file offload because it also carries the module-offload tests, the offloader-framework `vae` plumbing, and the MiniMax-H3 VAE-lifecycle guard; happy to split anything out if you'd prefer a narrower change.
